### PR TITLE
PRESIDECMS-651: Implemented clone form functionalities.

### PR DESCRIPTION
**Pull request by cfmitrah**

### CODE REVIEW

File: system/services/formbuilder/FormBuilderService.cfc [hunk 1]
```diff
@@ -959,6 +959,46 @@ component {
...
+		, required string name
+		, required string description
+	) {
+		var originalFormData = getForm( id=arguments.basedOnFormId );
+		var cloneFormData = structNew();
+
+		for( var column in originalFormData.columnList ) {
+			if( !listFindNoCase( "id,datecreated,datemodified,_version_is_draft,_version_has_drafts", column ) ) {
+				if( listFindNoCase( "name,description", column ) ) {
+					cloneFormData[ "#column#" ] = arguments[ "#column#" ];
```
Comment (DominicWatson) [inline system/services/formbuilder/FormBuilderService.cfc]: `cloneFormData[ "#column#" ]` should just be `cloneFormData[ column ]` (same for `arguments[ column ]`). See also lines below.

File: system/views/admin/formbuilder/cloneForm.cfm [hunk 2]
```diff
@@ -0,0 +1,28 @@
+<cfscript>
+	var id = rc.id ?: "";
+	var formId = createUUID();
```
Comment (DominicWatson) [inline system/views/admin/formbuilder/cloneForm.cfm]: This has the potential to create invalid HTML element IDs (they must not start with a number). I think ok to use "formbuilder-clone-form" or somesuch here.

File: system/services/formbuilder/FormBuilderService.cfc [hunk 3]
```diff
@@ -959,6 +959,46 @@ component {
 		}
 	}
 
+	public string function cloneForm(
+		  required string basedOnFormId
+		, required string name
+		, required string description
+	) {
+		var originalFormData = getForm( id=arguments.basedOnFormId );
+		var cloneFormData = structNew();
```
Comment (DominicWatson) [inline system/services/formbuilder/FormBuilderService.cfc]: Here, I think this is more accurate:

```
var cloneFormData = { name=arguments.name, description=arguments.description };
```

Then further down, when you loop over the `originalFormData.columnList`, you can ignore these fields and remove the logic on line 972-3.

File: system/services/formbuilder/FormBuilderService.cfc [hunk 4]
```diff
@@ -959,6 +959,46 @@ component {
...
+
+		for( var column in originalFormData.columnList ) {
+			if( !listFindNoCase( "id,datecreated,datemodified,_version_is_draft,_version_has_drafts", column ) ) {
+				if( listFindNoCase( "name,description", column ) ) {
```
Comment (DominicWatson) [inline system/services/formbuilder/FormBuilderService.cfc]: See comment above

**Review by DominicWatson — CHANGES_REQUESTED** [review]
Hey Mani, this is looking really great, thank you. Please see the review comments - a couple of minor tweaks, mostly just tidying.

Thanks!

**Comment by cfmitrah** [comment]
@DominicWatson  
	* Created valid HTML element IDs, using "formbuilder-clone-form".
	* Simplified the code while looping 'originalFormData.columnList'

File: system/views/admin/formbuilder/cloneForm.cfm (hunk 2 above)
Comment (cfmitrah) [reply system/views/admin/formbuilder/cloneForm.cfm]: Created valid HTML element IDs, using "formbuilder-clone-form".

File: system/services/formbuilder/FormBuilderService.cfc (hunk 3 above)
Comment (cfmitrah) [reply system/services/formbuilder/FormBuilderService.cfc]: Simplified the code while looping 'originalFormData.columnList' by 
<code>var cloneFormData = { name=arguments.name, description=arguments.description };</code>

**Comment by DominicWatson** [comment]
Super, thanks @cfmitrah. Merged into upcoming release.

**Pull request merged**